--- a/Makefile
+++ b/Makefile
@@ -1,10 +1,10 @@
-CXX=g++
-PREFIX=/usr/local
+CXX?=g++
+PREFIX=/usr
 
 F_VERSION=$(shell git log -n1 --date=short --format="%cd (commit=%h)" 2>/dev/null || echo "[unknown version - compile from git]")
 
-SNDFILE_INC?=$(shell pkg-config --cflags sndfile)
-SNDFILE_LIB?=$(shell pkg-config --libs sndfile)
+SNDFILE_INC=
+SNDFILE_LIB=-lsndfile
 
 CFLAGS=-D_FILE_OFFSET_BITS=64 -Wall -Wextra -W -Wno-unused-parameter -O3 -DFOLVE_VERSION='"$(F_VERSION)"' $(SNDFILE_INC)
 

