--- a/Makefile
+++ b/Makefile
@@ -40,10 +40,11 @@
 ######### Notify SDIO Host Keep Power During Syspend ##########
 CONFIG_RTW_SDIO_PM_KEEP_POWER = y
 ###################### Platform Related #######################
-CONFIG_PLATFORM_I386_PC = y
+CONFIG_PLATFORM_I386_PC = n
 ###############################################################
 
 CONFIG_DRVEXT_MODULE = n
+CONFIG_WIRELESS_EXT = y
 
 ########### COMMON  #################################
 
@@ -183,6 +184,10 @@
 
 ifeq ($(CONFIG_RTW_SDIO_PM_KEEP_POWER), y)
 EXTRA_CFLAGS += -DCONFIG_RTW_SDIO_PM_KEEP_POWER
+endif
+
+ifeq ($(CONFIG_WIRELESS_EXT), y)
+EXTRA_CFLAGS += -DCONFIG_WIRELESS_EXT
 endif
 
 ifeq ($(CONFIG_PLATFORM_I386_PC), y)
