--- a/drivers/net/wireless/rockchip_wlan/rkwifi/rk_wifi_config.c
+++ b/drivers/net/wireless/rockchip_wlan/rkwifi/rk_wifi_config.c
@@ -9,8 +9,8 @@
  *
  * gwl @ Rockchip
  */
-#include <linux/module.h>
 #include <linux/kernel.h>
+#include <linux/namei.h>
 #include <linux/rfkill-wlan.h>
 
 /* 
@@ -19,21 +19,35 @@
  
 #define VENDOR_ETC_FIRMWARE "/vendor/etc/firmware/"
 #define SYSTEM_ETC_FIRMWARE "/system/etc/firmware/"
+#define LINUX_LIB_FIRMWARE "/lib/firmware/"
+#define BROADCOM "brcm/"
 char ANDROID_FW_PATH[64] = {0};
 
 extern int get_wifi_chip_type(void);
 int rkwifi_set_firmware(char *fw, char *nvram)
 {
+    int err = 0;
+    struct path fw_dir;
     int chip = get_wifi_chip_type();
-    struct file *filp = NULL;
-
-    filp = filp_open(VENDOR_ETC_FIRMWARE, O_RDONLY, 0);
-
-    if (!IS_ERR(filp)) {
-	strcpy(ANDROID_FW_PATH, VENDOR_ETC_FIRMWARE);
+ 
+    err = kern_path(VENDOR_ETC_FIRMWARE, LOOKUP_DIRECTORY, &fw_dir);
+    if (! err) {
+        strcpy(ANDROID_FW_PATH, VENDOR_ETC_FIRMWARE);
     } else {
-	strcpy(ANDROID_FW_PATH, SYSTEM_ETC_FIRMWARE);
+        err = kern_path(SYSTEM_ETC_FIRMWARE, LOOKUP_DIRECTORY, &fw_dir);
+        if (! err) {
+            strcpy(ANDROID_FW_PATH, SYSTEM_ETC_FIRMWARE);
+        } else {
+            err = kern_path(LINUX_LIB_FIRMWARE, LOOKUP_DIRECTORY, &fw_dir);
+            if (! err) {
+                strcpy(ANDROID_FW_PATH, LINUX_LIB_FIRMWARE BROADCOM);
+            } else {
+                printk(KERN_WARNING "Rockchip firmware path undefined\n");
+                return 1;
+            }
+        }
     }
+    printk(KERN_INFO "Rockchip firmware path %s\n", ANDROID_FW_PATH);
 
 if (chip == WIFI_RK903) {
 	sprintf(fw, "%s%s", ANDROID_FW_PATH, "fw_RK903b2.bin");

