--- a/src/event/Thread.cxx
+++ b/src/event/Thread.cxx
@@ -47,6 +47,7 @@
 	event_loop.SetThread(ThreadId::GetCurrent());
 
 	if (realtime) {
++#ifndef ENABLE_RTOPT
 		SetThreadTimerSlack(std::chrono::microseconds(10));
 
 		try {
@@ -56,6 +57,7 @@
 				"RTIOThread could not get realtime scheduling, continuing anyway: {}",
 				std::current_exception());
 		}
++#endif # ENABLE_RTOPT
 	} else {
 #ifdef HAVE_URING
 		try {
