--- /home/iago/rockpi4/build/linux-custom/drivers/net/wireless/rockchip_wlan/rkwifi/bcmdhd/dhd_pno.c.orig
+++ /home/iago/rockpi4/build/linux-custom/drivers/net/wireless/rockchip_wlan/rkwifi/bcmdhd/dhd_pno.c
@@ -3165,7 +3165,11 @@
 	}
 	mutex_unlock(&_pno_state->pno_mutex);
 exit_no_unlock:
+#if defined CONFIG_PREEMPT
+	if (swait_active(&_pno_state->get_batch_done.wait))
+#else
 	if (waitqueue_active(&_pno_state->get_batch_done.wait))
+#endif
 		complete(&_pno_state->get_batch_done);
 	return err;
 }
@@ -3944,7 +3948,11 @@
 	{
 		struct dhd_pno_batch_params *params_batch;
 		params_batch = &_pno_state->pno_params_arr[INDEX_OF_BATCH_PARAMS].params_batch;
+#if defined CONFIG_PREEMPT
+		if (!swait_active(&_pno_state->get_batch_done.wait)) {
+#else
 		if (!waitqueue_active(&_pno_state->get_batch_done.wait)) {
+#endif
 			DHD_PNO(("%s : WLC_E_PFN_BEST_BATCHING\n", __FUNCTION__));
 			params_batch->get_batch.buf = NULL;
 			params_batch->get_batch.bufsize = 0;

