--- a/crypto/rand/rand_unix.c
+++ b/crypto/rand/rand_unix.c
@@ -476,6 +476,7 @@
     size_t entropy_available = 0;
     unsigned char *buffer;
 
+#undef OPENSSL_RAND_SEED_GETRANDOM
 #   if defined(OPENSSL_RAND_SEED_GETRANDOM)
     {
         ssize_t bytes;
@@ -510,6 +511,7 @@
     bytes_needed = rand_pool_bytes_needed(pool, 1 /*entropy_factor*/);
     {
         size_t i;
+#undef  DEVRANDOM_WAIT
 #ifdef DEVRANDOM_WAIT
         static int wait_done = 0;
 

