--- a/alsactl/info.c
+++ b/alsactl/info.c
@@ -21,6 +21,14 @@
 
 #include "aconfig.h"
 #include "alsactl.h"
+
+#ifndef HAVE_ALSA_PCM_H
+typedef int snd_pcm_stream_t;
+#endif
+
+#ifndef HAVE_ALSA_RAWMIDI_H
+typedef int snd_rawmidi_stream_t;
+#endif
 
 static int pcm_device_list(snd_ctl_t *ctl, snd_pcm_stream_t stream, bool *first)
 {
@@ -82,11 +90,13 @@
 
 static const char *snd_rawmidi_stream_name(snd_rawmidi_stream_t stream)
 {
+#ifdef __ALSA_RAWMIDI_H
 	if (stream == SND_RAWMIDI_STREAM_INPUT)
 		return "INPUT";
 	if (stream == SND_RAWMIDI_STREAM_OUTPUT)
 		return "OUTPUT";
 	return "???";
+#endif
 }
 
 static int rawmidi_device_list(snd_ctl_t *ctl, snd_rawmidi_stream_t stream, bool *first)

