--- a/aplay/aplay.c
+++ b/aplay/aplay.c
@@ -41,7 +41,9 @@
 #include <time.h>
 #include <locale.h>
 #include <alsa/asoundlib.h>
+#if defined HAVE_ALSA_USE_CASE_H
 #include <alsa/use-case.h>
+#endif
 #include <assert.h>
 #include <termios.h>
 #include <signal.h>
@@ -453,6 +455,7 @@
 	return offset;
 }
 
+#if defined HAVE_ALSA_USE_CASE_H
 static int open_ucm(snd_use_case_mgr_t **uc_mgr, char **pcm_name, const char *name)
 {
 	char *s, *p;
@@ -476,6 +479,7 @@
 	}
 	return err;
 }
+#endif
 
 static long parse_long(const char *str, int *err)
 {
@@ -553,7 +557,9 @@
 	int do_device_list = 0, do_pcm_list = 0, force_sample_format = 0;
 	snd_pcm_info_t *info;
 	FILE *direction;
+#if defined HAVE_ALSA_USE_CASE_H
 	snd_use_case_mgr_t *uc_mgr = NULL;
+#endif
 
 #ifdef ENABLE_NLS
 	setlocale(LC_ALL, "");
@@ -852,6 +858,7 @@
 		goto __end;
 	}
 
+#if defined HAVE_ALSA_USE_CASE_H
 	if (strncmp(pcm_name, "ucm.", 4) == 0) {
 		err = open_ucm(&uc_mgr, &pcm_name, pcm_name + 4);
 		if (err < 0) {
@@ -861,6 +868,7 @@
 		if (verbose)
 			fprintf(stderr, _("Found UCM PCM device: %s\n"), pcm_name);
 	}
+#endif
 
 	err = snd_pcm_open(&handle, pcm_name, stream, open_mode);
 	if (err < 0) {
@@ -951,8 +959,10 @@
 	if (verbose==2)
 		putchar('\n');
 	snd_pcm_close(handle);
+#if defined HAVE_ALSA_USE_CASE_H
 	if (uc_mgr)
 		snd_use_case_mgr_close(uc_mgr);
+#endif
 	handle = NULL;
 	free(audiobuf);
       __end:

