Fix albumart detection

Upstream: n/a (local customization)

Signed-off-by: Andreas Ziegler <15275159+aeolio@users.noreply.github.com>
---
--- a/src/command/FileCommands.cxx
+++ b/src/command/FileCommands.cxx
@@ -137,8 +137,9 @@
 find_stream_art(std::string_view directory, Mutex &mutex)
 {
 	static constexpr auto art_names = std::array {
+		"folder.jpg",
+		"cover.jpg",
 		"cover.png",
-		"cover.jpg",
 		"cover.webp",
 	};
 
