Adapt coremark to own environment
Upstream: configuration change
Signed-off-by: Andreas Ziegler <br025@umbiko.net>
---
--- a/linux/core_portme.h
+++ b/linux/core_portme.h
@@ -30,7 +30,7 @@
 	and implementation of functions thereof.
 */
 #ifndef USE_CLOCK
-#define USE_CLOCK 0
+#define USE_CLOCK 1
 #endif
 /* Configuration: HAS_STDIO
 	Define to 1 if the platform has stdio.h.
@@ -72,7 +72,7 @@
  #define COMPILER_FLAGS FLAGS_STR /* "Please put compiler flags here (e.g. -o3)" */
 #endif
 #ifndef MEM_LOCATION 
- #define MEM_LOCATION "Please put data memory location here\n\t\t\t(e.g. code in flash, data on heap etc)"
+ #define MEM_LOCATION "DDR DRAM"
  #define MEM_LOCATION_UNSPEC 1
 #endif
 

