--- brutefir.orig/bfio_alsa.c
+++ brutefir-1.0o/bfio_alsa.c
@@ -446,7 +446,8 @@
 	memcpy(fd2as[pollfd.fd].channel_selection, channel_selection,
 	       used_channels * sizeof(int));
     }
-    efree(settings->device);
+    if (settings->device != NULL)
+        efree(settings->device);
     efree(settings);
     return pollfd.fd;
 }
@@ -595,6 +596,15 @@
     const uint8_t *ptr;
     int n, i, err;
 
+    if (as->restart) {
+        as->restart = false;
+        if ((err = snd_pcm_prepare(as->handle)) < 0) {
+            fprintf(stderr, "ALSA I/O: Could not restart audio: %s.\n",
+                    snd_strerror(err));
+            errno = -EPIPE;
+            return -1;
+        }
+    }
     if (as->isinterleaved) {
 	i = as->sample_size * as->open_channels;
 	if ((n = snd_pcm_writei(as->handle, &((const uint8_t *)buf)[offset],
@@ -614,15 +624,6 @@
             goto bfio_write_error;
         }
     }
-    if (as->restart) {
-        as->restart = false;
-        if ((err = snd_pcm_start(as->handle)) < 0) {
-            fprintf(stderr, "ALSA I/O: Could not restart audio: %s.\n",
-                    snd_strerror(err));
-            errno = -EPIPE;
-            return -1;
-        }
-    }
     return n * i;
 
  bfio_write_error:
