--- a/meson.build
+++ b/meson.build
@@ -195,6 +195,9 @@
 conf.set('ENABLE_INOTIFY', enable_inotify)
 
 conf.set('ENABLE_DSD', get_option('dsd'))
+
+enable_rtopt = get_option('rtopt')
+conf.set('ENABLE_RTOPT', enable_rtopt)
 
 inc = include_directories(
   'src',
@@ -316,6 +319,12 @@
   'src/PlaylistFile.cxx',
 ]
 
+if enable_rtopt
+  sources += [
+    'src/thread/RealtimeOption.cxx',
+  ]
+endif
+
 if not is_android
   sources += [
     'src/CommandLine.cxx',
@@ -475,6 +484,12 @@
 else
   target_type = 'executable'
   target_name = 'mpd'
+endif
+
+if compiler.find_library('libiconv', required: false).found()
+  link_args += [
+    '-liconv',
+  ]
 endif
 
 mpd = build_target(
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -193,3 +193,5 @@
        choices: ['auto', 'avahi', 'bonjour', 'disabled'],
        value: 'auto',
        description: 'Zeroconf support')
+
+option('rtopt', type: 'boolean', value: false, description: 'Yan\'s RT patch')
