--- a/src/config/Option.hxx
+++ b/src/config/Option.hxx
@@ -20,6 +20,7 @@
 #ifndef MPD_CONFIG_OPTION_HXX
 #define MPD_CONFIG_OPTION_HXX
 
+#include <config.h>
 #include "util/Compiler.h"
 
 #if defined(_WIN32) && CLANG_OR_GCC_VERSION(4,7)
@@ -91,6 +92,9 @@
 	RESAMPLER,
 	AUDIO_FILTER,
 	DATABASE,
+#ifdef ENABLE_RTOPT
+        REALTIME_OPTION,
+#endif
 	NEIGHBORS,
 	MAX
 };
--- a/src/config/Templates.cxx
+++ b/src/config/Templates.cxx
@@ -22,6 +22,7 @@
 
 #include <iterator>
 
+#include <config.h>
 #include <string.h>
 
 const ConfigTemplate config_param_templates[] = {
@@ -92,6 +93,9 @@
 	{ "resampler" },
 	{ "filter", true },
 	{ "database" },
+#ifdef ENABLE_RTOPT
+        { "realtime_option", false },
+#endif
 	{ "neighbors", true },
 };
 
