diff -Naur a/src/sched/sched_getparam.c b/musl-1.1.19/src/sched/sched_getparam.c
--- a/src/sched/sched_getparam.c	2018-02-22 19:39:19.000000000 +0100
+++ b/src/sched/sched_getparam.c	2018-06-28 12:03:58.636142791 +0200
@@ -1,8 +1,7 @@
 #include <sched.h>
-#include <errno.h>
 #include "syscall.h"
 
 int sched_getparam(pid_t pid, struct sched_param *param)
 {
-	return __syscall_ret(-ENOSYS);
+	return syscall(SYS_sched_getparam, pid, param);
 }
diff -Naur a/src/sched/sched_getscheduler.c b/src/sched/sched_getscheduler.c
--- a/src/sched/sched_getscheduler.c	2018-02-22 19:39:19.000000000 +0100
+++ b/src/sched/sched_getscheduler.c	2018-06-28 12:05:16.140589375 +0200
@@ -1,8 +1,7 @@
 #include <sched.h>
-#include <errno.h>
 #include "syscall.h"
 
 int sched_getscheduler(pid_t pid)
 {
-	return __syscall_ret(-ENOSYS);
+	return syscall(SYS_sched_getscheduler, pid);
 }
diff -Naur a/src/sched/sched_setparam.c b/src/sched/sched_setparam.c
--- a/src/sched/sched_setparam.c	2018-02-22 19:39:19.000000000 +0100
+++ b/src/sched/sched_setparam.c	2018-06-28 12:06:42.468802065 +0200
@@ -1,8 +1,7 @@
 #include <sched.h>
-#include <errno.h>
 #include "syscall.h"
 
 int sched_setparam(pid_t pid, const struct sched_param *param)
 {
-	return __syscall_ret(-ENOSYS);
+	syscall(SYS_sched_setparam, pid, param);
 }
diff -Naur a/src/sched/sched_setscheduler.c b/src/sched/sched_setscheduler.c
--- a/src/sched/sched_setscheduler.c	2018-02-22 19:39:19.000000000 +0100
+++ b/src/sched/sched_setscheduler.c	2018-06-28 12:07:38.388835523 +0200
@@ -1,8 +1,7 @@
 #include <sched.h>
-#include <errno.h>
 #include "syscall.h"
 
 int sched_setscheduler(pid_t pid, int sched, const struct sched_param *param)
 {
-	return __syscall_ret(-ENOSYS);
+	return syscall(SYS_sched_setscheduler, pid, sched, param);
 }
