--- a/doc/config/lighttpd.conf
+++ b/doc/config/lighttpd.conf
@@ -16,7 +16,7 @@
 var.log_root    = "/var/log"
 var.server_root = "/var/www"
 var.state_dir   = "/var/run"
-var.home_dir    = "/var/lib/lighttpd"
+var.home_dir    = "/usr/lib/lighttpd"
 var.conf_dir    = "/etc/lighttpd"
 
 ## 
@@ -355,7 +355,7 @@
 ##
 ## directory listing configuration
 ##
-include "conf.d/dirlisting.conf"
+#include "conf.d/dirlisting.conf"
 
 ##
 ## Should lighttpd follow symlinks?
