--- a/sound/soc/codecs/rk3308_codec_provider.h
+++ b/sound/soc/codecs/rk3308_codec_provider.h
@@ -20,7 +20,7 @@
 #ifndef __RK3308_CODEC_PROVIDER_H__
 #define __RK3308_CODEC_PROVIDER_H__
 
-#ifdef CONFIG_SND_SOC_RK3308
+#if defined(CONFIG_SND_SOC_RK3308) || defined(CONFIG_SND_SOC_RK3308_MODULE)
 void rk3308_codec_set_jack_detect(struct snd_soc_codec *codec,
 				  struct snd_soc_jack *hpdet_jack);
 #else

